src/PrestaShopBundle/EventListener/ModuleActivatedListener.php

```diff
@@ -0,0 +1,208 @@
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\EventListener;
+
+use Doctrine\Common\Annotations\AnnotationException;
+use PrestaShopBundle\Security\Annotation\ModuleActivated;
+use PrestaShop\PrestaShop\Core\Addon\Module\ModuleRepository;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\HttpKernel\Event\FilterControllerEvent;
+use Symfony\Component\Translation\TranslatorInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\Routing\RouterInterface;
+use Doctrine\Common\Annotations\Reader;
+use Doctrine\Common\Util\ClassUtils;
+use ReflectionObject;
+use ReflectionClass;
+use Module;
+
+/**
+ * Allow a redirection to the right url when using ModuleActivated annotation
+ * and the module is inactive.
+ */
+class ModuleActivatedListener
+{
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var SessionInterface
+     */
+    private $session;
+
+    /**
+     * @var Reader
+     */
+    private $annotationReader;
+
+    /**
+     * @var ModuleRepository
+     */
+    private $moduleRepository;
+
+    /**
+     * @param RouterInterface $router
+     * @param TranslatorInterface $translator
+     * @param SessionInterface $session
+     * @param Reader $annotationReader
+     * @param ModuleRepository $moduleRepository
+     */
+    public function __construct(
+        RouterInterface $router,
+        TranslatorInterface $translator,
+        SessionInterface $session,
+        Reader $annotationReader,
+        ModuleRepository $moduleRepository
+    ) {
+        $this->router = $router;
+        $this->translator = $translator;
+        $this->session = $session;
+        $this->annotationReader = $annotationReader;
+        $this->moduleRepository = $moduleRepository;
+    }
+
+    /**
+     * @param FilterControllerEvent $event
+     *
+     * @throws AnnotationException
+     * @throws \ReflectionException
+     */
+    public function onKernelController(FilterControllerEvent $event)
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        $controller = $event->getController();
+
+        if (!is_array($controller)) {
+            return;
+        }
+
+        list($controllerObject, $methodName) = $controller;
+        $moduleActivated = $this->getAnnotation($controllerObject, $methodName);
+
+        if (null === $moduleActivated) {
+            return;
+        }
+
+        /** @var Module $module */
+        $module = $this->moduleRepository->getInstanceByName($moduleActivated->getModuleName());
```

Comment:
Thinking about it, it's not good enough (sorry!) because it defeats my purpose: what I want is no call at all to object models in the bundle: here we have just hidden and moved the object model.

How about the relying on the Module class from Adapter namespace instead? Or best of it (and @eternoendless will love that) create some kind of ModuleFeature using FeatureInterface from Adapter and DBAL connection to only retrieve if the module is activated or not?